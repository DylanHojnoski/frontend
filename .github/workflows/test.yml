name: Test

on:
  - push
  - pull_request

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/setup-node@v2
        with:
          node-version: '16'

      - uses: actions/checkout@v1

      - name: Get yarn cache directory path
        id: yarn-cache-dir-path
        run: echo "::set-output name=dir::$(yarn cache dir)"

      - name: Cache yarn dependencies
        uses: actions/cache@v2
        id: yarn-cache
        with:
          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            ${{ runner.os }}-yarn-

      - name: Install dependencies
        run: yarn install

      - name: Run tests & get coverage
        run: yarn test:coverage

      - name: Upload coverage
        uses: codecov/codecov-action@v1
        with:
          files: ./coverage/lcov.info
