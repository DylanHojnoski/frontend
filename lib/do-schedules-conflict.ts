import {Schedule} from './rschedule';

const getCommonElementsInArrays = <T>(array1: T[], array2: T[]) => {
	const common: T[] = [];

	for (const element of array1) {
		if (array2.includes(element)) {
			common.push(element);
		}
	}

	return common;
};

/** Compare two times.
 * Returns 0 if equal, -1 if first time is earlier, 1 if first time is later
 */
const compareTimes = (time1: Date, time2: Date) => {
	const hours1 = time1.getHours();
	const minutes1 = time1.getMinutes();
	const hours2 = time2.getHours();
	const minutes2 = time2.getMinutes();

	const totalMinutes1 = (hours1 * 60) + minutes1;
	const totalMinutes2 = (hours2 * 60) + minutes2;

	const diff = totalMinutes1 - totalMinutes2;

	if (diff === 0) {
		return 0;
	}

	if (diff > 0) {
		return 1;
	}

	return -1;
};

const doSchedulesConflict = (firstSchedule: Schedule, secondSchedule: Schedule) => {
	// There's a much more elegant solution to this using the intersection operator from rSchedule.
	// However, static analysis is far faster.
	// See: https://gitlab.com/john.carroll.p/rschedule/-/issues/61

	// Quick & cheap check
	if (firstSchedule.firstDate?.date.getTime() === secondSchedule.firstDate?.date.getTime()) {
		return true;
	}

	const firstRuleSet = firstSchedule.rrules;
	const secondRuleSet = secondSchedule.rrules;

	// In the vast majority of cases there will be a single rule in each set,
	// so O(n^2) shouldn't matter too much. (Famous last words...)
	for (const currentFirstRuleSetRule of firstRuleSet) {
		for (const currentSecondRuleSetRule of secondRuleSet) {
			if (!currentFirstRuleSetRule.firstDate?.date || !currentFirstRuleSetRule.lastDate?.date) {
				continue;
			}

			if (!currentSecondRuleSetRule.firstDate?.date || !currentSecondRuleSetRule.lastDate?.date) {
				continue;
			}

			// If rules occur during completely separate calendar periods
			if (currentFirstRuleSetRule.lastDate.date < currentSecondRuleSetRule.firstDate.date ||
          currentSecondRuleSetRule.lastDate.date < currentFirstRuleSetRule.firstDate.date) {
				continue;
			}

			const firstByDayOfWeek = currentFirstRuleSetRule.options.byDayOfWeek;
			const secondByDayOfWeek = currentSecondRuleSetRule.options.byDayOfWeek;

			if (!firstByDayOfWeek || !secondByDayOfWeek) {
				continue;
			}

			const firstStart = new Date(currentFirstRuleSetRule.options.start.valueOf());
			const firstEnd = new Date(currentFirstRuleSetRule.options.end?.valueOf() ?? 0);
			const secondStart = new Date(currentSecondRuleSetRule.options.start.valueOf());
			const secondEnd = new Date(currentSecondRuleSetRule.options.end?.valueOf() ?? 0);

			if (getCommonElementsInArrays(firstByDayOfWeek, secondByDayOfWeek).length > 0) {
				const compareStartResult = compareTimes(firstStart, secondStart);

				if (compareStartResult === 0) {
					return true;
				}

				if (compareStartResult === -1) { // Check if first starts before second
					// Check if end overlaps
					if (compareTimes(firstEnd, secondStart) === 1) {
						return true;
					}
				} else if (compareStartResult === 1 && // Check if first starts after second
          // Check if start overlaps
          compareTimes(firstEnd, secondEnd) === -1) {
					return true;
				}

				return true;
			}
		}
	}

	return false;
};

export default doSchedulesConflict;
