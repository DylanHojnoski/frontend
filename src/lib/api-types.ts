import {type Except} from 'type-fest';
import {type Schedule} from './rschedule';

export enum ESemester {
	SPRING = 'SPRING',
	FALL = 'FALL',
	SUMMER = 'SUMMER',
}

export enum ELocationType {
	PHYSICAL = 'PHYSICAL',
	ONLINE = 'ONLINE',
	REMOTE = 'REMOTE',
	UNKNOWN = 'UNKNOWN',
}

export type IInstructorFromAPI = {
	id: number;
	fullName: string;
	departments: string[];
	email: string | null;
	phone: string | null;
	office: string | null;
	websiteURL: string | null;
	photoURL: string | null;
	thumbnailURL: string | null;
	interests: string[];
	occupations: string[];
	averageDifficultyRating: number | null;
	averageRating: number | null;
	numRatings: number | null;
	rmpId: string | null;
	updatedAt: string;
	deletedAt: string | null;
};

export type IPassFailDropRecord = {
	year: number;
	dropped: number;
	failed: number;
	total: number;
	semester: ESemester;
};

export type IPassFailDropFromAPI = Record<string, IPassFailDropRecord[]>;

export type IBuildingFromAPI = {
	name: string;
	shortName: string;
	lat: number;
	lon: number;
};

export type ISectionFromAPI = {
	id: string;
	courseId: string;
	crn: string;
	section: string;
	cmp: string;
	minCredits: number;
	maxCredits: number;
	time: Schedule.JSON;
	totalSeats: number;
	takenSeats: number;
	availableSeats: number;
	fee: number;
	instructors: Array<{
		id: number;
	}>;
	locationType: ELocationType;
	buildingName: string | null;
	room: string | null;
	updatedAt: string;
	deletedAt: string | null;
};

export type ISectionFromAPIWithSchedule = ISectionFromAPI & {
	parsedTime: Schedule | null;
};

export type ICourseFromAPI = {
	id: string;
	year: number;
	semester: ESemester;
	subject: string;
	crse: string;
	title: string;
	description: string | null;
	prereqs: string | null;
	updatedAt: string;
	deletedAt: string | null;
	offered: string[] | null;
	minCredits: number;
	maxCredits: number;
};
export type ITransferCourseFromAPI = {
	id: string;
	fromCollege: string;
	fromCollegeState: string;
	fromCRSE: string;
	fromCredits: number;
	fromSubject: string;
	toCRSE: string;
	toCredits: number;
	toSubject: string;
	title: string;
	updatedAt: string;
};

export type IFullCourseFromAPI = ICourseFromAPI & {
	sections: Array<Except<ISectionFromAPI, 'instructors'> &
	{instructors: Array<Except<IInstructorFromAPI, 'thumbnailURL'>>}>;
};
