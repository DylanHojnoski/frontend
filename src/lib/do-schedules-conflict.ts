import compareTimes from './compare-times';
import {Schedule} from './rschedule';

const getCommonElementsInArrays = <T>(array1: T[], array2: T[]) => {
	const common: T[] = [];

	for (const element of array1) {
		if (array2.includes(element)) {
			common.push(element);
		}
	}

	return common;
};

const addDuration = (date: Date, ms: number) => new Date(date.getTime() + ms);

const doSchedulesConflict = (firstSchedule: Schedule, secondSchedule: Schedule) => {
	// There's a much more elegant solution to this using the intersection operator from rSchedule.
	// However, static analysis is far faster.
	// See: https://gitlab.com/john.carroll.p/rschedule/-/issues/61

	// Quick & cheap check
	if (firstSchedule.firstDate?.date.getTime() === secondSchedule.firstDate?.date.getTime()) {
		return true;
	}

	const firstRuleSet = firstSchedule.rrules;
	const secondRuleSet = secondSchedule.rrules;

	// In the vast majority of cases there will be a single rule in each set,
	// so O(n^2) shouldn't matter too much. (Famous last words...)
	for (const currentFirstRuleSetRule of firstRuleSet) {
		for (const currentSecondRuleSetRule of secondRuleSet) {
			if (!currentFirstRuleSetRule.firstDate?.date || !currentFirstRuleSetRule.lastDate?.date) {
				continue;
			}

			if (!currentSecondRuleSetRule.firstDate?.date || !currentSecondRuleSetRule.lastDate?.date) {
				continue;
			}

			// If rules occur during completely separate calendar periods
			if (currentFirstRuleSetRule.lastDate.date < currentSecondRuleSetRule.firstDate.date
          || currentSecondRuleSetRule.lastDate.date < currentFirstRuleSetRule.firstDate.date) {
				continue;
			}

			const firstByDayOfWeek = currentFirstRuleSetRule.options.byDayOfWeek;
			const secondByDayOfWeek = currentSecondRuleSetRule.options.byDayOfWeek;

			if (!firstByDayOfWeek || !secondByDayOfWeek) {
				continue;
			}

			const firstStart = currentFirstRuleSetRule.firstDate.date;
			const firstEnd = addDuration(firstStart, currentFirstRuleSetRule.duration!);
			const secondStart = currentSecondRuleSetRule.firstDate.date;
			const secondEnd = addDuration(secondStart, currentSecondRuleSetRule.duration!);

			if (getCommonElementsInArrays(firstByDayOfWeek, secondByDayOfWeek).length > 0) {
				const compareStartResult = compareTimes(firstStart, secondStart);

				if (compareStartResult === 0) {
					return true;
				}

				if (compareStartResult === -1) { // Check if first starts before second
					// Check if end overlaps
					if (compareTimes(firstEnd, secondStart) === 1) {
						return true;
					}
				} else if (compareStartResult === 1 // Check if first starts after second
          // Check if start overlaps
          && compareTimes(firstStart, secondEnd) === -1) {
					return true;
				}
			}
		}
	}

	return false;
};

export default doSchedulesConflict;
