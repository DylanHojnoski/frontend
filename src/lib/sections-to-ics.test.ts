import {ISectionFromAPI, ICourseFromAPI, ESemester, ELocationType} from './api-types';
import {THRICE_WEEKLY} from './do-schedules-conflict.test';
import sectionsToICS from './sections-to-ics';

const COURSE: ICourseFromAPI = {
	id: 'test-course-id',
	year: 2020,
	semester: ESemester.FALL,
	subject: 'CS',
	crse: '1000',
	title: 'Intro to Programming',
	description: '',
	prereqs: null,
	updatedAt: new Date().toISOString(),
	deletedAt: null,
	minCredits: 3,
	maxCredits: 3,
	offered: [],
};

const SECTION: ISectionFromAPI & {course: ICourseFromAPI} = {
	course: COURSE,

	id: 'test-section-id',
	courseId: COURSE.id,
	crn: 'test-crn',
	section: '1A',
	cmp: '1',
	minCredits: 3,
	maxCredits: 3,
	totalSeats: 20,
	takenSeats: 10,
	availableSeats: 10,
	fee: 0,
	instructors: [],
	locationType: ELocationType.PHYSICAL,
	buildingName: null,
	room: null,
	updatedAt: new Date().toISOString(),
	deletedAt: null,

	time: THRICE_WEEKLY.toJSON(),
};

test('snapshot', () => {
	let calendarString = sectionsToICS([SECTION], []);

	// Remove non-deterministic timestamp
	calendarString = calendarString.replace(/DTSTAMP:[^\n]+\n/, '');
	// Remove rrule, slightly different depending on timezone but
	// we don't care because the day is important, not the time
	calendarString = calendarString.replaceAll(/RRULE:[^\n]+\n/g, '');
	expect(calendarString).toMatchSnapshot();
});
