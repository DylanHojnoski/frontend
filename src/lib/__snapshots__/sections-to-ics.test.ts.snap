// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//sebbo.net//ical-generator//EN
BEGIN:VTIMEZONE
TZID:America/New_York
TZURL:http://tzurl.org/zoneinfo-outlook/America/New_York
X-LIC-LOCATION:America/New_York
BEGIN:DAYLIGHT
TZOFFSETFROM:-0500
TZOFFSETTO:-0400
TZNAME:EDT
DTSTART:19700308T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=2SU
END:DAYLIGHT
BEGIN:STANDARD
TZOFFSETFROM:-0400
TZOFFSETTO:-0500
TZNAME:EST
DTSTART:19701101T020000
RRULE:FREQ=YEARLY;BYMONTH=11;BYDAY=1SU
END:STANDARD
END:VTIMEZONE
TIMEZONE-ID:America/New_York
X-WR-TIMEZONE:America/New_York
BEGIN:VEVENT
UID:test-section-id
SEQUENCE:0
DTSTART;TZID=America/New_York:20210702T100000
DTEND;TZID=America/New_York:20210702T110000
RRULE:FREQ=WEEKLY;UNTIL=20211230T090000;BYDAY=MO,WE,FR
SUMMARY:CS1000 (Intro to Programming)
DESCRIPTION:
BEGIN:VALARM
ACTION:DISPLAY
TRIGGER:-PT10M
DESCRIPTION:CS1000 (Intro to Programming)
END:VALARM
END:VEVENT
END:VCALENDAR"
`;
